internal/app/export/service.go

```diff
@@ -4,62 +4,294 @@
+	<-e.connected
+	e.Logger.Info("Starting republish, waiting for stream data")
+	for _, route := range e.Cfg.Routes {
+		streams := []string{route.NatsTopic, ">"}
+		go func() {
+			for {
+				msgs, err := e.readMessages(streams)
+				if err != nil {
+					continue
+				}
+				e.Logger.Info(fmt.Sprintf("Waiting for connection to %s", e.Cfg.MQTT.Host))
+				for {
+					// Wait for connection
+					if e.IsConnected() || <-e.connected {
+						for _, m := range msgs {
+							if err := e.publish(m.Topic, []byte(m.Payload)); err != nil {
+								e.Logger.Error("Failed to republish message")
+							}
+						}
+						break
+					}
+				}
+			}
+		}()
+	}
+}
+
+func (e *exporter) readMessages(streams []string) (map[string]messages.Msg, error) {
+	// Wait for messages in cache, blocking read
+	msgs, read, err := e.Cache.ReadGroup(streams, exportGroup, count, e.ID)
+
+	switch err {
+	case messages.ErrDecodingData:
+		e.Logger.Error(fmt.Sprintf("Failed to decode all data from stream. Read: %d, Failed: %d, Batch: %d.", len(msgs), read, count))
+	default:
+		e.Logger.Error(fmt.Sprintf("Failed to read from stream %s", err.Error()))
+		return nil, err
+	}
+	e.Logger.Info(fmt.Sprintf("Read %d records from the stream", len(msgs)))
+	return msgs, nil
+}
+
+func (e *exporter) Subscribe(topic string, nc *nats.Conn) {
+	_, err := nc.QueueSubscribe(topic, e.ID, e.Consume)
 	if err != nil {
-		e.Logger.Error(fmt.Sprintf("Failed to subscribe for NATS/MQTT %s/%s", r.NatsTopic(), r.MqttTopic()))
+		e.Logger.Error(fmt.Sprintf("Failed to subscribe for NATS %s", topic))
+	}
+}
+
+func (e *exporter) publish(topic string, payload []byte) error {
+	if e.connectionStatus() != connected {
+		e.Logger.Error("not connected to mqtt broker")
+		return mqtt.ErrNotConnected
+	}
+	token := e.MQTT.Publish(topic, byte(e.Cfg.MQTT.QoS), e.Cfg.MQTT.Retain, payload)
+	if token.Wait() && token.Error() != nil {
+		e.Logger.Error(fmt.Sprintf("Failed to publish to topic %s", topic))
+		return token.Error()
+	}
+	return nil
+}
+
+func (e *exporter) validateSubject(sub string) bool {
+	if strings.ContainsAny(sub, " \t\r\n") {
+		return false
+	}
+	tokens := strings.Split(sub, ".")
+	for _, t := range tokens {
+		if len(t) == 0 {
+			return false
+		}
+	}
+	return true
+}
+
+func (e *exporter) conn(client mqtt.Client) {
+	e.setConnected(connected)
+	e.Logger.Debug(fmt.Sprintf("Client %s connected", e.ID))
+}
+
+func (e *exporter) lost(client mqtt.Client, err error) {
+	e.setConnected(disconnected)
+	e.Logger.Debug(fmt.Sprintf("Client %s disconnected", e.ID))
+}
+
+// IsConnected returns a bool signifying whether the client is connected or not.
+func (e *exporter) IsConnected() bool {
+	e.RLock()
+	defer e.RUnlock()
+	status := atomic.LoadUint32(&e.status)
+	switch {
+	case status == connected:
+		return true
+	default:
+		return false
+	}
+}
+
+func (e *exporter) connectionStatus() uint32 {
+	e.RLock()
+	defer e.RUnlock()
+	status := atomic.LoadUint32(&e.status)
```

Comment:
You don't need both Lock and Atomic.

Reply:
I've copied that solution from mqtt lib

Reply:
Are you sure that they know what they are doing in MQTT lib? Did you check? Paho lib is pretty badly wrtitten.

Reply:
Are you sure that they know what they are doing in MQTT lib? Did you check? Paho lib is pretty badly wrtitten.

Reply:
You can set `--race` flag to validate the solution when testing.